### PR TITLE
allow setting different certificates for kube-controller-managed CSR signers

### PR DESCRIPTION
**Pull request by deads2k**
builds on https://github.com/kubernetes/kubernetes/pull/90731

The kube-controller-manager managed signers can now have distinct signing certificates and keys.  See the help about `--cluster-signing-[signer-name]-{cert,key}-file`.  `--cluster-signing-{cert,key}-file` is still the default.

/kind feature
/priority important-soon
@kubernetes/sig-auth-pr-reviews 

```release-note
The kube-controller-manager managed signers can now have distinct signing certificates and keys.  See the help about `--cluster-signing-[signer-name]-{cert,key}-file`.  `--cluster-signing-{cert,key}-file` is still the default.
```

### CODE REVIEW

File: pkg/controller/certificates/signer/config/types.go [hunk 1]
```diff
@@ -28,7 +28,26 @@ type CSRSigningControllerConfiguration struct {
 	// clusterSigningCertFile is the filename containing a PEM-encoded
 	// RSA or ECDSA private key used to issue cluster-scoped certificates
 	ClusterSigningKeyFile string
+
+	// kubeletServingSignerConfiguration holds the certificate and key used to issue certificates for the kubernetes.io/kubelet-serving signer
+	KubeletServingSignerConfiguration CSRSigningConfiguration
+	// kubeletClientSignerConfiguration holds the certificate and key used to issue certificates for the kubernetes.io/kube-apiserver-client-kubelet
+	KubeletClientSignerConfiguration CSRSigningConfiguration
```
Comment (munnerz) [inline pkg/controller/certificates/signer/config/types.go:35]: Do we intend/want to support having a distinct CA for kube-apiserver-client and kube-apiserver-client-kubelet? Presumably these would be reasonably expected to be two intermediaries from a common root (once there is better support for intermediaries too).

Unless it’s possible to specify multiple root CAs used to validate apiserver client certificates?

File: pkg/controller/certificates/signer/config/types.go [hunk 2]
```diff
@@ -28,7 +28,26 @@ type CSRSigningControllerConfiguration struct {
...
 	// will be given.
 	ClusterSigningDuration metav1.Duration
 }
+
+type CSRSigningConfiguration struct {
+	// certFile is the filename containing a PEM-encoded
+	// X509 CA certificate used to issue certificates
+	CertFile string
+	// keyFile is the filename containing a PEM-encoded
+	// RSA or ECDSA private key used to issue certificates
```
Comment (munnerz) [inline pkg/controller/certificates/signer/config/types.go:51]: What formats are supported here? PKCS#8? What about PKCS#1/whatever the equivalent of 1 is for ECDSA?

File: pkg/controller/certificates/signer/config/types.go (hunk 1 above)
Comment (deads2k) [reply pkg/controller/certificates/signer/config/types.go:35]: > Unless it’s possible to specify multiple root CAs used to validate apiserver client certificates?

It is.  In fact openshift does this today.



> Do we intend/want to support having a distinct CA for kube-apiserver-client and kube-apiserver-client-kubelet?

Yes.  The kubelets have well defined and well behaved rotation capabilities that are managed, not external.  These properties mean that having comparatively short lived signers is possible.  In general, apiserver clients don't have these properties.

File: pkg/controller/certificates/signer/config/types.go (hunk 2 above)
Comment (deads2k) [reply pkg/controller/certificates/signer/config/types.go:51]: > What formats are supported here? PKCS#8? What about PKCS#1/whatever the equivalent of 1 is for ECDSA?

This is equivalent to the doc on the existing field.  I'll admit to cargo-culting.

File: api/api-rules/violation_exceptions.list [hunk 3]
```diff
@@ -482,9 +482,15 @@ API rule violation: names_match,k8s.io/apimachinery/pkg/util/intstr,IntOrString,
 API rule violation: names_match,k8s.io/apimachinery/pkg/util/intstr,IntOrString,Type
 API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,AttachDetachControllerConfiguration,DisableAttachDetachReconcilerSync
 API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,AttachDetachControllerConfiguration,ReconcilerSyncLoopPeriod
+API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,CSRSigningConfiguration,CertFile
+API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,CSRSigningConfiguration,KeyFile
 API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,CSRSigningControllerConfiguration,ClusterSigningCertFile
 API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,CSRSigningControllerConfiguration,ClusterSigningDuration
 API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,CSRSigningControllerConfiguration,ClusterSigningKeyFile
+API rule violation: names_match,k8s.io/kube-controller-manager/config/v1alpha1,CSRSigningControllerConfiguration,KubeAPIServerClientSignerConfiguration
```
Comment (deads2k) [inline api/api-rules/violation_exceptions.list:490]: Notice that the other fields in this struct are here too.  JSON tags haven't been added to these types at all yet.

**Comment by micahhausler** [comment]
Maybe a dumb question, but could I use this to turn off specific signers?  If I only wanted to sign node-serving-cert but not node-client cert, could I do that with this change?

**Comment by enj** [comment]
> Maybe a dumb question, but could I use this to turn off specific signers? If I only wanted to sign node-serving-cert but not node-client cert, could I do that with this change?

No, but you could set the `node-client-cert` to point to a private key that is not trusted anywhere, which would make it inert.  Not really a solution though ...

File: cmd/kube-controller-manager/app/certificates.go [hunk 4]
```diff
@@ -44,7 +44,12 @@ func startCSRSigningController(ctx ControllerContext) (http.Handler, bool, error
...
+	hasSingleSigningFile := ctx.ComponentConfig.CSRSigningController.ClusterSigningCertFile == "" || ctx.ComponentConfig.CSRSigningController.ClusterSigningKeyFile == ""
+	hasKubeletServingFile := ctx.ComponentConfig.CSRSigningController.KubeletServingSignerConfiguration.CertFile == "" || ctx.ComponentConfig.CSRSigningController.KubeletServingSignerConfiguration.KeyFile == ""
+	hasKubeletClientFile := ctx.ComponentConfig.CSRSigningController.KubeletClientSignerConfiguration.CertFile == "" || ctx.ComponentConfig.CSRSigningController.KubeletClientSignerConfiguration.KeyFile == ""
+	hasKubeAPIServerClientFile := ctx.ComponentConfig.CSRSigningController.KubeAPIServerClientSignerConfiguration.CertFile == "" || ctx.ComponentConfig.CSRSigningController.KubeAPIServerClientSignerConfiguration.KeyFile == ""
+	hasLegacyUnknownFile := ctx.ComponentConfig.CSRSigningController.LegacyUnknownSignerConfiguration.CertFile == "" || ctx.ComponentConfig.CSRSigningController.LegacyUnknownSignerConfiguration.KeyFile == ""
+	if !hasSingleSigningFile && !hasKubeletServingFile && !hasKubeletClientFile && !hasKubeAPIServerClientFile && !hasLegacyUnknownFile {
```
Comment (enj) [inline cmd/kube-controller-manager/app/certificates.go:52]: Er, what happens if you specify one of the signers but not the default signing file?

Can we add some tests to exercise this wiring?

**Review by enj — COMMENTED** [review]
Overall LGTM.

I feel like we should add something like a `--cluster-signing-explicit-signers` flag which is a `bool` (default `false`) that, when set to `true`, prevents KCM from starting when each signer does not have an explicit key/cert specified.  This would allow you to know that a new signer has been added and that you need to address that in some way.  Otherwise a new signer addition would implicitly use the default cert/key file, which may not be desired.

File: cmd/kube-controller-manager/app/certificates.go (hunk 4 above)
Comment (deads2k) [reply cmd/kube-controller-manager/app/certificates.go:52]: > Er, what happens if you specify one of the signers but not the default signing file?

You'll get an error about an unspecified file.  Comes from the code doing the initial load.  What is your preferred behavior?

Comment (deads2k) [reply cmd/kube-controller-manager/app/certificates.go:52]: > Can we add some tests to exercise this wiring?

Do you have an example of flag wiring behavior tests you like?  I don't object, I just haven't seen many.

Comment (enj) [reply cmd/kube-controller-manager/app/certificates.go:52]: > You'll get an error about an unspecified file. Comes from the code doing the initial load. What is your preferred behavior?

That seems fine to me.

> Do you have an example of flag wiring behavior tests you like?

There are some tests in `kube-controller-manager/app/options/options_test.go` that test the input flags to `CSRSigningControllerOptions` bits.  Maybe combine that with `startCSRSigningController` and some permutations of the flags?

File: cmd/kube-controller-manager/app/certificates.go [hunk 5]
```diff
@@ -44,7 +44,12 @@ func startCSRSigningController(ctx ControllerContext) (http.Handler, bool, error
...
 		return nil, false, nil
 	}
-	if ctx.ComponentConfig.CSRSigningController.ClusterSigningCertFile == "" || ctx.ComponentConfig.CSRSigningController.ClusterSigningKeyFile == "" {
+	hasSingleSigningFile := ctx.ComponentConfig.CSRSigningController.ClusterSigningCertFile == "" || ctx.ComponentConfig.CSRSigningController.ClusterSigningKeyFile == ""
+	hasKubeletServingFile := ctx.ComponentConfig.CSRSigningController.KubeletServingSignerConfiguration.CertFile == "" || ctx.ComponentConfig.CSRSigningController.KubeletServingSignerConfiguration.KeyFile == ""
+	hasKubeletClientFile := ctx.ComponentConfig.CSRSigningController.KubeletClientSignerConfiguration.CertFile == "" || ctx.ComponentConfig.CSRSigningController.KubeletClientSignerConfiguration.KeyFile == ""
+	hasKubeAPIServerClientFile := ctx.ComponentConfig.CSRSigningController.KubeAPIServerClientSignerConfiguration.CertFile == "" || ctx.ComponentConfig.CSRSigningController.KubeAPIServerClientSignerConfiguration.KeyFile == ""
+	hasLegacyUnknownFile := ctx.ComponentConfig.CSRSigningController.LegacyUnknownSignerConfiguration.CertFile == "" || ctx.ComponentConfig.CSRSigningController.LegacyUnknownSignerConfiguration.KeyFile == ""
```
Comment (liggitt) [inline cmd/kube-controller-manager/app/certificates.go:51]: ow, my eyes... maybe

```go
csrConfig := ctx.ComponentConfig.CSRSigningController
hasSingleSigningFile := csrConfig.ClusterSigningCertFile...
...
```
?

Comment (liggitt) [reply cmd/kube-controller-manager/app/certificates.go:51]: also, are all of these right? shouldn't these be checking if either of the params are *not* empty?

File: cmd/kube-controller-manager/app/certificates.go [hunk 6]
```diff
@@ -71,6 +76,9 @@ func startCSRSigningController(ctx ControllerContext) (http.Handler, bool, error
 	keyUsesDefault = (ctx.ComponentConfig.CSRSigningController.ClusterSigningKeyFile == kubeoptions.DefaultClusterSigningKeyFile)
 
 	switch {
+	case anySpecificFilesSet(ctx.ComponentConfig.CSRSigningController):
+		// in this case we continue because the default keys will not be used. Their status doesn't matter anymore.
```
Comment (liggitt) [inline cmd/kube-controller-manager/app/certificates.go:80]: does this reflect what we discussed last week about disallowing combined use of the general ca/key files and specific signer loop ca/key files?

Comment (deads2k) [reply cmd/kube-controller-manager/app/certificates.go:80]: > does this reflect what we discussed last week about disallowing combined use of the general ca/key files and specific signer loop ca/key files?

Yes it does.

Comment (deads2k) [reply cmd/kube-controller-manager/app/certificates.go:80]: > does this reflect what we discussed last week about disallowing combined use of the general ca/key files and specific signer loop ca/key files?

correct.  Specifying a cert/key pair for any specific loop will prevent the general (old flag) from being used.

File: cmd/kube-controller-manager/app/certificates.go (hunk 5 above)
Comment (deads2k) [reply cmd/kube-controller-manager/app/certificates.go]: fixed names and updated to use the helper with a correctly named function.

File: cmd/kube-controller-manager/app/certificates.go [hunk 7]
```diff
@@ -44,7 +44,8 @@ func startCSRSigningController(ctx ControllerContext) (http.Handler, bool, error
 		klog.Warningf("Resource %s is not available now", gvr.String())
 		return nil, false, nil
 	}
-	if ctx.ComponentConfig.CSRSigningController.ClusterSigningCertFile == "" || ctx.ComponentConfig.CSRSigningController.ClusterSigningKeyFile == "" {
+	missingSingleSigningFile := ctx.ComponentConfig.CSRSigningController.ClusterSigningCertFile == "" || ctx.ComponentConfig.CSRSigningController.ClusterSigningKeyFile == ""
+	if missingSingleSigningFile && !anySpecificFilesSet(ctx.ComponentConfig.CSRSigningController) {
```
Comment (enj) [inline cmd/kube-controller-manager/app/certificates.go:48]: We need an extra error case like:

```go
if !missingSingleSigningFile && anySpecificFilesSet(ctx.ComponentConfig.CSRSigningController) {
     return nil, false, fmt.Errorf("cannot specify default and per controller certs at the same time")
}
```

right?

**Review by enj — CHANGES_REQUESTED** [review]

File: cmd/kube-controller-manager/app/certificates.go (hunk 6 above)
Comment (liggitt) [reply cmd/kube-controller-manager/app/certificates.go:76]: would this be a good time to remove the default values for the general flags (as we've been indicating we will in the warning below) and error here if they set both the general and specific flags?

**Comment by deads2k** [comment]
pushed commit to refine the help text and address comments.

File: cmd/kube-controller-manager/app/options/csrsigningcontroller.go [hunk 8]
```diff
@@ -70,5 +84,34 @@ func (o *CSRSigningControllerOptions) Validate() []error {
 	}
 
 	errs := []error{}
```
Comment (liggitt) [inline cmd/kube-controller-manager/app/options/csrsigningcontroller.go:86]: is this the right place to ensure:
* general and individual files are not both provided
* `(len(o.ClusterSigningCertFile) == 0) == (len(o.ClusterSigningKeyFile) == 0)`

File: cmd/kube-controller-manager/app/certificates.go (hunk 6 above)
Comment (liggitt) [reply cmd/kube-controller-manager/app/certificates.go:79]: (would remove cert/key file defaults in RecommendedDefaultCSRSigningControllerConfiguration and this entire switch statement)

File: cmd/kube-controller-manager/app/options/csrsigningcontroller.go (hunk 8 above)
Comment (deads2k) [reply cmd/kube-controller-manager/app/options/csrsigningcontroller.go:86]: > is this the right place to ensure:
> 
> * general and individual files are not both provided
> * `(len(o.ClusterSigningCertFile) == 0) == (len(o.ClusterSigningKeyFile) == 0)`

happy to also specify here if you like.

Comment (deads2k) [reply cmd/kube-controller-manager/app/options/csrsigningcontroller.go:86]: > is this the right place to ensure:
> 
> * general and individual files are not both provided
> * `(len(o.ClusterSigningCertFile) == 0) == (len(o.ClusterSigningKeyFile) == 0)`

added

**Comment by enj** [comment]
/lgtm
/retest

Please squash before merge.

**Comment by deads2k** [comment]
squashed.

**Comment by deads2k** [comment]
/retest

I think I've seen these verify timeouts as being temporary.

**Pull request merged**